### PR TITLE
Day2str

### PR DESCRIPTION
**Pull request by alejandro-colomar**
- I've moved my refactors out of <https://github.com/shadow-maint/shadow/pull/876>, since the first two commits are more important, and might be good before 4.15, but these aren't so much.

### CODE REVIEW

**Comment by alejandro-colomar** [comment]
v2 changes:

-  Use lib/time/ subdir

```
$ git range-diff master gh/day2str day2str 
1:  741fc096 = 1:  741fc096 src/chage.c: Unify long overflow checks in print_day_as_date()
2:  d5d9194e = 2:  d5d9194e lib/: Saturate addition to avoid overflow
3:  20bb258f < -:  -------- lib/, src/, po/: date_to_str(): Move function to header, and make inline
-:  -------- > 3:  3bd301e4 lib/, src/, po/: date_to_str(): Move function to header, and make inline
4:  46362e76 ! 4:  b636afcf lib/day_to_str.h: DAY_TO_STR(): Add macro
    @@ Metadata
     Author: Alejandro Colomar <alx@kernel.org>
     
      ## Commit message ##
    -    lib/day_to_str.h: DAY_TO_STR(): Add macro
    +    lib/time/day_to_str.h: DAY_TO_STR(): Add macro
     
         This macro ensures that the buffer is an array, and calculates the size.
     
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
    - ## lib/day_to_str.h ##
    + ## lib/time/day_to_str.h ##
     @@
      
      #include <time.h>
5:  fd9ca675 = 5:  7416616a src/: Use DAY_TO_STR() instead of its pattern
6:  ce086de9 = 6:  47e450c3 src/passwd.c: print_status(): Fix typo (bogus use of the comma operator)
7:  1a73b4d1 ! 7:  f77a1143 lib/day_to_str.[ch]: day_to_str(): Accept a day instead of a date, and rename function
    @@ Metadata
     Author: Alejandro Colomar <alx@kernel.org>
     
      ## Commit message ##
    -    lib/day_to_str.[ch]: day_to_str(): Accept a day instead of a date, and rename function
    +    lib/time/day_to_str.[ch]: day_to_str(): Accept a day instead of a date, and rename function
     
         It was always being called with 'day * DAY', so do that internally and
         simplify.  This grabs some code from print_day_as_date().
    @@ Commit message
         Cc: Tobias Stoeckmann <tobias@stoeckmann.org>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
    - ## lib/day_to_str.c ##
    + ## lib/time/day_to_str.c ##
     @@
      // SPDX-FileCopyrightText: 2021-2024, Alejandro Colomar <alx@kernel.org>
     +// SPDX-FileCopyrightText: 2024, Tobias Stoeckmann <tobias@stoeckmann.org>
    @@ lib/day_to_str.c
      
      
     @@
    - #include "day_to_str.h"
    + #include "time/day_to_str.h"
      
      
     -extern inline void date_to_str(size_t size, char buf[size], long date);
     +extern inline void day_to_str(size_t size, char buf[size], long day);
     
    - ## lib/day_to_str.h ##
    + ## lib/time/day_to_str.h ##
     @@
     -/*
     - * SPDX-FileCopyrightText: 2021-2023, Alejandro Colomar <alx@kernel.org>
    @@ lib/day_to_str.h
     +// SPDX-License-Identifier: BSD-3-Clause
      
      
    - #ifndef SHADOW_INCLUDE_LIB_DAY_TO_STR_H_
    + #ifndef SHADOW_INCLUDE_LIB_TIME_DAY_TO_STR_H_
     @@
      #include "string/strtcpy.h"
      
8:  d7b2729e ! 8:  e058ebdc lib/, src/: Call gmtime_r(3) instead of gmtime(3)
    @@ Commit message
     
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
    - ## lib/day_to_str.h ##
    -@@ lib/day_to_str.h: inline void day_to_str(size_t size, char buf[size], long day);
    + ## lib/time/day_to_str.h ##
    +@@ lib/time/day_to_str.h: inline void day_to_str(size_t size, char buf[size], long day);
      inline void
      day_to_str(size_t size, char buf[size], long day)
      {
    @@ lib/day_to_str.h: inline void day_to_str(size_t size, char buf[size], long day);
      
        if (day < 0) {
                strtcpy(buf, "never", size);
    -@@ lib/day_to_str.h: day_to_str(size_t size, char buf[size], long day)
    +@@ lib/time/day_to_str.h: day_to_str(size_t size, char buf[size], long day)
                return;
        }
      
```

**Comment by alejandro-colomar** [comment]
v3 changes:

-  Rebase to master

```
$ git range-diff alx/master..gh/day2str shadow/master..day2str 
1:  741fc096 < -:  -------- src/chage.c: Unify long overflow checks in print_day_as_date()
2:  d5d9194e < -:  -------- lib/: Saturate addition to avoid overflow
3:  3bd301e4 = 1:  2c05a34f lib/, src/, po/: date_to_str(): Move function to header, and make inline
4:  b636afcf = 2:  16c2bcb9 lib/time/day_to_str.h: DAY_TO_STR(): Add macro
5:  7416616a = 3:  e8da3365 src/: Use DAY_TO_STR() instead of its pattern
6:  47e450c3 = 4:  77a1b196 src/passwd.c: print_status(): Fix typo (bogus use of the comma operator)
7:  f77a1143 = 5:  540711ee lib/time/day_to_str.[ch]: day_to_str(): Accept a day instead of a date, and rename function
8:  e058ebdc = 6:  3c07d2dd lib/, src/: Call gmtime_r(3) instead of gmtime(3)
```

**Comment by alejandro-colomar** [comment]
v3b changes:

-  Rebase to master

```
$ git range-diff gh/master..gh/day2str alx/master..day2str 
1:  2c05a34f = 1:  c4cdcdf5 lib/, src/, po/: date_to_str(): Move function to header, and make inline
2:  16c2bcb9 = 2:  3f86cf8e lib/time/day_to_str.h: DAY_TO_STR(): Add macro
3:  e8da3365 = 3:  ac9eba28 src/: Use DAY_TO_STR() instead of its pattern
4:  77a1b196 = 4:  38707801 src/passwd.c: print_status(): Fix typo (bogus use of the comma operator)
5:  540711ee = 5:  80af7302 lib/time/day_to_str.[ch]: day_to_str(): Accept a day instead of a date, and rename function
6:  3c07d2dd = 6:  739dada8 lib/, src/: Call gmtime_r(3) instead of gmtime(3)
```

**Review by hallyn — APPROVED** [review]

File: lib/time/day_to_str.h [hunk 1]
```diff
@@ -17,32 +16,36 @@
...
-	tm = gmtime(&t);
+	if (__builtin_mul_overflow(day, DAY, &date)) {
+		strtcpy(buf, "future", size);
+		return;
+	}
+
+	tm = gmtime(&date);
 	if (tm == NULL) {
-		(void) strtcpy(buf, "future", size);
+		strtcpy(buf, "future", size);
```
Comment (hallyn) [inline lib/time/day_to_str.h:42]: I realize that there are no callers who could get hit by this, however if buf were too short for "future", then buf would be truncated and not null-terminated.

This seems a bad precedent to set.  It seems like, unless the caller is prepared for this, we should check strtcpy return value and set buf[size-1] = 0 if truncation occurred.

I've merged this pr, my comment here is a more general question.

Comment (alejandro-colomar) [reply lib/time/day_to_str.h:42]: strtcpy() NULL terminates short copies.  It's almost equivalent to the kernel's strscpy(9).

Comment (alejandro-colomar) [reply lib/time/day_to_str.h:42]:
```c
$ grepc -tfd strtcpy .
./lib/string/strtcpy.h:inline ssize_t
strtcpy(char *restrict dst, const char *restrict src, size_t dsize)
{
	bool    trunc;
	size_t  dlen, slen;

	if (dsize == 0)
		return -1;

	slen = strnlen(src, dsize);
	trunc = (slen == dsize);
	dlen = slen - trunc;

	stpcpy(mempcpy(dst, src, dlen), "");

	if (trunc)
		return -1;

	return slen;
}
```

The main work is here:
```c
stpcpy(mempcpy(dst, src, dlen), "");
```

And that `stpcpy(..., "");` ensures null termination.

Comment (hallyn) [reply lib/time/day_to_str.h:42]: Ah, thanks.

I was inverting the value of @trunc in my head.

Looks good.

**Pull request merged**